drivers/azure/azureutil/securityrules.go

```diff
@@ -0,0 +1,313 @@
+			Description:              to.StringPtr("Allow ssh from public Internet"),
+			SourceAddressPrefix:      to.StringPtr("*"),
+			DestinationAddressPrefix: to.StringPtr("*"),
+			SourcePortRange:          to.StringPtr("*"),
+			DestinationPortRange:     to.StringPtr(fmt.Sprintf("%d", sshPort)),
+			Access:                   network.SecurityRuleAccessAllow,
+			Direction:                network.SecurityRuleDirectionInbound,
+			Protocol:                 network.SecurityRuleProtocolTCP,
+			Priority:                 to.Int32Ptr(sshPriority),
+		},
+	})
+	return nil
+}
+
+func addDockerRule(rules *[]network.SecurityRule, dockerPort int) error {
+	// Check if any existing rule has the same priority as the Docker rule or check any conflicts
+	for _, rule := range *rules {
+		if strings.ToLower(fmt.Sprintf("%s", rule.Direction)) != "inbound" {
+			// Ignore outbound rules
+			continue
+		}
+		priority := to.Int32(rule.Priority)
+		if priority == dockerPriority {
+			name := to.String(rule.Name)
+			if name == dockerPolicyName {
+				// Docker rule already exists
+				return nil
+			}
+			return fmt.Errorf("Cannot set Docker rule as priority %d is already taken by rule %s", dockerPriority, name)
+		}
+		if priority < dockerPriority {
+			// If the priority is less than the docker priority, we should check if there's a conflict
+			err := checkIfDockerConflict(rule, dockerPort)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	// If not, add the rule
+	*rules = append(*rules, network.SecurityRule{
+		Name: to.StringPtr(dockerPolicyName),
+		SecurityRulePropertiesFormat: &network.SecurityRulePropertiesFormat{
+			Description:              to.StringPtr("Allow docker engine access (TLS-protected)"),
+			SourceAddressPrefix:      to.StringPtr("*"),
+			DestinationAddressPrefix: to.StringPtr("*"),
+			SourcePortRange:          to.StringPtr("*"),
+			DestinationPortRange:     to.StringPtr(fmt.Sprintf("%d", dockerPort)),
+			Access:                   network.SecurityRuleAccessAllow,
+			Direction:                network.SecurityRuleDirectionInbound,
+			Protocol:                 network.SecurityRuleProtocolTCP,
+			Priority:                 to.Int32Ptr(dockerPriority),
+		},
+	})
+	return nil
+}
+
+func syncOpenPorts(rules *[]network.SecurityRule, openPorts []string) error {
+	openPortsToBeDeleted := make(map[string]network.SecurityRule)
+	takenPriorities := make(map[int32]bool)
+	// Add all rules for deletion by default
+	for _, rule := range *rules {
+		takenPriorities[to.Int32(rule.Priority)] = true
+		if isOpenPort(rule) {
+			openPortsToBeDeleted[to.String(rule.Name)] = rule
+		}
+	}
+	// Remove from deletion if it exists or add to queue if it does not exist
+	openPortsToBeAdded := make(map[string]network.SecurityRule)
+	for _, p := range openPorts {
+		op := parseOpenPort(p)
+		if op == nil {
+			return fmt.Errorf("unable to parse open port %s", p)
+		}
+		if _, ok := openPortsToBeDeleted[op.name]; ok {
+			// Do not delete the open port
```

Comment:
nit: comment is a little confusing since it's followed by delete, maybe "remove port from list of open ports to be deleted"

Reply:
updated!